base/rational.jl

```diff
@@ -129,12 +129,220 @@ Bool(x::Rational) = x==0 ? false : x==1 ? true :
 (::Type{T})(x::Rational) where {T<:Integer} = (isinteger(x) ? convert(T, x.num)::T :
     throw(InexactError(nameof(T), T, x)))
 
-AbstractFloat(x::Rational) = (float(x.num)/float(x.den))::AbstractFloat
-function (::Type{T})(x::Rational{S}) where T<:AbstractFloat where S
-    P = promote_type(T,S)
-    convert(T, convert(P,x.num)/convert(P,x.den))::T
+bit_width(n) = ndigits(n, base = UInt8(2), pad = false)
+
+function divrem_pow2(num::Integer, n::Integer)
+    quot = num >> n
+    rema = num - (quot << n)
+    (quot, rema)
+end
+
+const RoundingModesIndependentFromSign = Union{
+    RoundingMode{:ToZero}, RoundingMode{:FromZero},
+    RoundingMode{:Nearest}, RoundingMode{:NearestTiesAway}}
+
+rounding_mode_translated_for_abs(rm::RoundingModesIndependentFromSign, ::Real) = rm
+
+function rounding_mode_translated_for_abs(::RoundingMode{:Down}, sign::Real)
+    !signbit(sign) ? RoundToZero : RoundFromZero
+end
+
+function rounding_mode_translated_for_abs(::RoundingMode{:Up}, sign::Real)
+    !signbit(sign) ? RoundFromZero : RoundToZero
+end
+
+clamped_to_zero(x) = max(zero(x), x)
+
+struct FloatComponentsResult{S<:Integer,T<:Integer}
+    mantissa::S
+    exponent::T
+    inexact::Bool
+    underflowed::Bool
+
+    function FloatComponentsResult(m::S, e::T;
+                                   inexact::Bool,
+                                   underflowed::Bool) where {S<:Integer,T<:Integer}
+        new{S,T}(m, e, inexact, underflowed)
+    end
+end
+
+# `num`, `den` are positive integers. `requested_precision` is the
+# requested floating-point precision. `T` is the integer type that
+# we'll be working with mostly, it needs to be wide enough.
+# `exp_lower_bound` is the minimum allowed normalized exponent for
+# normal numbers.
+function rational_to_float_components_impl(num::Integer,
+                                           den::Integer,
+                                           requested_precision::Integer,
+                                           ::Type{T},
+                                           romo::RoundingMode,
+                                           exp_lower_bound::Union{Nothing,Integer}) where
+{T<:Integer}
```

Comment:
```suggestion
                                           exp_lower_bound::Union{Nothing,Integer}) where {T<:Integer}
```

It's fine to put that on the same line.